core/src/packets/icmp/v6/ndp/mod.rs

```diff
@@ -19,105 +19,797 @@
+/// add custom NDP options. This trait should not be imported and used
+/// directly.
+pub trait NdpOption<'a> {
+    /// Returns the option type.
+    fn option_type(&self) -> NdpOptionType;
+
+    /// Returns the length of the option in units of 8 octets.
+    fn length(&self) -> u8;
+
+    /// Parses the buffer at offset as this NDP option.
+    ///
+    /// The buffer offset includes `option_type` and `length` fields. The
+    /// implementation should verify that `option_type` matches the expected
+    /// type code. Otherwise parse should fail. When applicable, the
+    /// implementation should also verify that `length` is consistent with
+    /// the expected length for the given option type as well.
+    ///
+    /// # Remarks
+    ///
+    /// This function cannot be invoked directly. It is internally used by
+    /// [`ImmutableNdpOption::downcast`] and [`MutableNdpOption::downcast`].
+    ///
+    /// [`ImmutableNdpOption::downcast`]: ImmutableNdpOption::downcast
+    /// [`MutableNdpOption::downcast`]: MutableNdpOption::downcast
+    fn try_parse(mbuf: &'a mut Mbuf, offset: usize, internal: Internal) -> Fallible<Self>
+    where
+        Self: Sized;
+
+    /// Pushes a new NDP option to the buffer at offset.
+    ///
+    /// # Remarks
+    ///
+    /// This function cannot be invoked directly. It is internally used by
+    /// [`NdpOptions::prepend`] and [`NdpOptions::append`].
+    ///
+    /// [`NdpOptions::prepend`]: NdpOptions::prepend
+    /// [`NdpOptions::append`]: NdpOptions::append
+    fn try_push(mbuf: &'a mut Mbuf, offset: usize, internal: Internal) -> Fallible<Self>
+    where
+        Self: Sized;
+}
+
 #[cfg(test)]
 mod tests {
     use super::*;
-    use crate::net::MacAddr;
-    use crate::packets::icmp::v6::ndp::NdpOptions;
     use crate::packets::ip::v6::Ipv6;
-    use crate::packets::{Ethernet, Packet};
+    use crate::packets::Ethernet;
     use crate::testils::byte_arrays::ROUTER_ADVERT_PACKET;
-    use crate::Mbuf;
-    use fallible_iterator::FallibleIterator;
-    use std::str::FromStr;
 
     #[capsule::test]
-    fn find_source_link_layer_address() {
+    fn iterate_immutable_ndp_options() {
+        let packet = Mbuf::from_bytes(&ROUTER_ADVERT_PACKET).unwrap();
+        let ethernet = packet.parse::<Ethernet>().unwrap();
+        let ipv6 = ethernet.parse::<Ipv6>().unwrap();
+        let advert = ipv6.parse::<RouterAdvertisement<Ipv6>>().unwrap();
+
+        let mut prefix = false;
+        let mut mtu = false;
+        let mut source = false;
+        let mut other = false;
+
+        let mut iter = advert.options_iter();
+
+        while let Some(option) = iter.next().unwrap() {
+            match option.option_type() {
+                NdpOptionTypes::PrefixInformation => prefix = true,
+                NdpOptionTypes::Mtu => mtu = true,
+                NdpOptionTypes::SourceLinkLayerAddress => source = true,
+                _ => other = true,
```

Comment:
minor: do we notate anywhere, interior comment, that this other is an option we don't currently handle? Maybe just for posterity?